unison-cli/src/Unison/Codebase/Editor/Slurp.hs

```diff
@@ -52,61 +51,44 @@ untagged (TypeVar v) = v
-  deriving (Eq, Ord, Show)
-
--- | A definition's final status, incorporating the statuses of all of its dependencies.
-data SummarizedStatus v
-  = SelfStatus DefnStatus
-  | -- Dependency Status
-    DepStatus (TaggedVar v) DefnStatus
-  deriving (Eq, Ord, Show)
-
--- | Ideally we would display all available information about each var to the end-user,
--- but for now we simply pick the "most important" issue to maintain backwards compatibility
--- with current behaviour.
-pickPriorityStatus :: SummarizedStatus v -> SummarizedStatus v -> SummarizedStatus v
-pickPriorityStatus a b =
-  case (a, b) of
-    -- If the definition has its own error, that takes highest priority.
-    (SelfStatus (DefErr err), _) -> SelfStatus (DefErr err)
-    (_, SelfStatus (DefErr err)) -> SelfStatus (DefErr err)
-    -- Next we care if a dependency has an error
-    (DepStatus v (DefErr err), _) -> DepStatus v (DefErr err)
-    (_, DepStatus v (DefErr err)) -> DepStatus v (DefErr err)
-    -- If our definition needs its own update then we don't care if dependencies need updates.
-    (SelfStatus (DefOk Updated), _) -> SelfStatus (DefOk Updated)
-    (_, SelfStatus (DefOk Updated)) -> SelfStatus (DefOk Updated)
-    (DepStatus v (DefOk Updated), _) -> DepStatus v (DefOk Updated)
-    (_, DepStatus v (DefOk Updated)) -> DepStatus v (DefOk Updated)
-    -- Any other 'ok' dependency status doesn't meaningfully affect anything the summary.
-    (DepStatus _ (DefOk _), x) -> x
-    (x, DepStatus _ (DefOk _)) -> x
-    -- 'New' definitions take precedence over duplicated dependencies when reporting status.
-    -- E.g. if a definition has dependencies which are duplicated, but it is itself a new
-    -- definition, we report it as New.
-    (SelfStatus (DefOk New), _) -> SelfStatus (DefOk New)
-    (_, SelfStatus (DefOk New)) -> SelfStatus (DefOk New)
-    (SelfStatus (DefOk Duplicated), _) -> SelfStatus (DefOk Duplicated)
+-- | A coarser, totally-ordered variant of a defnintion's status, which summarizes its own status and the statuses of
+-- all of its transitive dependencies.
+--
+-- For example, if any transitive dependency of a defnition requires an `update`, then so does the definition itself,
+-- even if it's new (and thus ok to `add`).
+data DepStatus
+  = -- | Part of a term/ctor or ctor/term collision: neither `add` nor `update` ok
+    DepCollision
+  | -- | Requires an update: `add` not ok, `update` ok
+    DepNeedsUpdate
+  | -- | `add` or `update` both ok
+    DepOk
+  deriving stock (Eq, Ord, Show)
+
+-- | Classify a definition status into a coarser dependency status.
+defnStatusToDepStatus :: DefnStatus -> DepStatus
+defnStatusToDepStatus = \case
+  CtorTermCollision -> DepCollision
+  Duplicated -> DepOk
+  New -> DepOk
+  TermCtorCollision -> DepCollision
+  Updated -> DepNeedsUpdate
+
+-- | DepCollision more severe than DepNeedsUpdate more severe than DepOk
+mostSevereDepStatus :: DepStatus -> DepStatus -> DepStatus
+mostSevereDepStatus =
+  min
```

Comment:
I'd personally prefer an explicit `case` here just to guard against new constructors being added in the wrong spot or w/e; If not that, then a note on the constructors indicating that they're defined in 'severity' order.

Reply:
I'll add a note to the docs; I think an explicit `case` might still have the same issue, unless we actually list out every single combination. And if we do list out every single combination, someone will eventually try to clean it up one day with underscores, which brings us back to something with the same pitfalls as `min`.